src/JWT.ts

```diff
@@ -386,12 +376,34 @@ export async function createMultisignatureJWT(
 }
 
 export function verifyJWSDecoded(
-  { header, data, signature }: JWSDecoded,
+  { header, payload, data, signature }: JWTDecoded,
   pubKeys: VerificationMethod | VerificationMethod[]
 ): VerificationMethod {
   if (!Array.isArray(pubKeys)) pubKeys = [pubKeys]
-  const signer: VerificationMethod = VerifierAlgorithm(header.alg)(data, signature, pubKeys)
-  return signer
+
+  const iss = payload.iss
+  let level = 0
+  let recurse = true
+  do {
+    console.log(`verifyJWSDecoded(): checking JWT at level ${level}`)
+    if (iss !== payload.iss) throw new Error(`${JWT_ERROR.INVALID_JWT}: multiple issuers`)
+
+    try {
+      return VerifierAlgorithm(header.alg)(data, signature, pubKeys)
+    } catch (e) {
+      if (!(e as Error).message.startsWith(JWT_ERROR.INVALID_SIGNATURE)) throw e
+    }
+
+    // TODO probably best to create copy objects than replace reference objects
+    if (header.cty !== 'JWT') {
+      recurse = false
+    } else {
+      ;({ payload, header, signature, data } = decodeJWT(payload.jwt, false))
+    }
+    level++
+  } while (recurse)
+
+  throw new Error(`${JWT_ERROR.INVALID_SIGNATURE}: no matching public key found`)
```

Comment:
won't the loop be stopped after the return? what is the function of this loop?

Reply:
L392 `return VerifierAlgorithm(header.alg)(data, signature, pubKeys)` throws an Error if the provided signature does not match and the loop continues. If each of the nested JWT levels do not have a matching signature then it will get to L406 and through and `INVALID_SIGNATURE` error. Please confirm this makes sense to you?

Reply:
i understand thank you